File: app/src/main/java/com/example/mcabr/detectsensors/MainActivity.java
package com.example.mcabr.detectsensors;

import android.app.Activity;
import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorManager;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends Activity {

        ListView listViewSensor;
        private ArrayList<Sensor> foundSensor = new ArrayList<Sensor>();
        private ArrayAdapter<Sensor> aaSensor;

        private SensorManager sensorManager;

        public void onCreate(Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
            setContentView(R.layout.activity_main);

            initUI();

            getSensorManager();

            getSensorList();
        }

    public void initUI(){
        listViewSensor = (ListView) findViewById(R.id.listViewSensor);

        aaSensor = new ArrayAdapter<Sensor>(this,
                android.R.layout.simple_list_item_1, foundSensor);

        listViewSensor.setAdapter(aaSensor);
        listViewSensor.setOnItemClickListener(myListSensorClickListener);
    }

    public void getSensorManager(){
        // TODO obtener en la variable sensorManager una instancia del servicio SENSOR_SERVICE
        sensorManager = (SensorManager)getSystemService(Context.SENSOR_SERVICE);
    }

    public String getTypeSensor(int type){
        String cadType;

        // TODO tipos de sensores de entero a cadena de texto
        // EJ. Sensor.TYPE_ACCELEROMETER --> cadType = "ACCELEROMETER";

        switch (type) {
            case (Sensor.TYPE_ACCELEROMETER) :
                cadType = "ACCELEROMETER";
                break;
            case (Sensor.TYPE_GRAVITY) :
                cadType = "GRAVITY";
                break;
            case (Sensor.TYPE_GYROSCOPE) :
                cadType = "GRAVITY";
                break;
            case (Sensor.TYPE_LIGHT) :
                cadType = "LIGHT";
                break;
            case (Sensor.TYPE_LINEAR_ACCELERATION) :
                cadType = "LINEAR_ACCELERATION";
                break;
            case (Sensor.TYPE_MAGNETIC_FIELD) :
                cadType = "MAGNETIC_FIELD";
                break;
            case (Sensor.TYPE_ORIENTATION) : // Deprecated
                cadType = "ORIENTATION";
                break;
            case (Sensor.TYPE_PRESSURE) :
                cadType = "PRESSURE";
                break;
            case (Sensor.TYPE_PROXIMITY) :
                cadType = "PROXIMITY";
                break;
            case (Sensor.TYPE_ROTATION_VECTOR) :
                cadType = "ROTATION_VECTOR";
                break;
            case (Sensor.TYPE_AMBIENT_TEMPERATURE) :
                cadType = "TEMPERATURE";
                break;
            default:
                cadType = "OTHER";
                break;
        }
        return cadType;
    }

    public void getSensorList(){
        // TODO Obtener la lista de todos los sensores
        List<Sensor> sensors = sensorManager.getSensorList(Sensor.TYPE_ALL);

        for (Sensor sensor : sensors) {
            if (!foundSensor.contains(sensor))
                foundSensor.add(sensor);
        }
        aaSensor.notifyDataSetChanged();
    }

    AdapterView.OnItemClickListener myListSensorClickListener = new AdapterView.OnItemClickListener() {
        public void onItemClick(AdapterView<?> arg0, View view, int index,
                                long arg3) {

            // TODO sensor que ha sido seleccionado en la lista
            Sensor mySensor = aaSensor.getItem(index);

            // TODO informacion sobre el sensor seleccionado
            // Por ejemplo, su tipo y su nombre
            // Podeis usar la funcion getTypeSensor para obtener el tipo del sensor en String
            String infoSensor = getTypeSensor(mySensor.getType()) + " - " + mySensor.getName();

            // TODO Imprimir en pantalla
            Toast.makeText(getApplicationContext(), infoSensor, Toast.LENGTH_LONG).show();
        }
    };

}

File: app/src/main/java/com/example/mcabr/detectsensors/MainActivity.kt
package com.example.mcabr.detectsensors

import android.app.Activity
import android.app.AlertDialog
import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorManager
import android.os.Bundle
import android.view.View
import android.widget.*

class MainActivity : Activity() {

    private lateinit var listViewSensor: ListView
    private val sensorList = mutableListOf<Sensor>()
    private var aaSensor: SensorAdapter? = null

    private var sensorManager: SensorManager? = null

    private var myListSensorClickListener: AdapterView.OnItemClickListener = AdapterView.OnItemClickListener { _, _, index, _ ->
        val mySensor = aaSensor!!.getItem(index)

        val dialogBuilder = AlertDialog.Builder(this)
        val dialogView = View.inflate(this, R.layout.sensor_layout,null)
        dialogBuilder.setView(dialogView)

        val name = dialogView.findViewById<TextView>(R.id.name)
        val vendor = dialogView.findViewById<TextView>(R.id.vendor)
        val version = dialogView.findViewById<TextView>(R.id.version)
        val type = dialogView.findViewById<TextView>(R.id.type)
        val maxRange = dialogView.findViewById<TextView>(R.id.maxRange)
        val resolution = dialogView.findViewById<TextView>(R.id.resolution)
        val power = dialogView.findViewById<TextView>(R.id.power)
        val minDelay = dialogView.findViewById<TextView>(R.id.minDelay)

        name.text = mySensor.name
        vendor.text = mySensor.vendor
        version.text = mySensor.version.toString()
        type.text = mySensor.stringType
        maxRange.text = mySensor.maximumRange.toString()
        resolution.text = mySensor.resolution.toString()
        power.text = mySensor.power.toString()
        minDelay.text = mySensor.minDelay.toString()

        dialogBuilder.create().show()
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        initUI()
        getSensorManager()
        getSensorList()
    }

    private fun initUI() {
        listViewSensor = findViewById<View>(R.id.listViewSensor) as ListView

        aaSensor = SensorAdapter(this,
                 sensorList)

        listViewSensor.adapter = aaSensor
        listViewSensor.onItemClickListener = myListSensorClickListener
    }

    private fun getSensorManager() {
        sensorManager = getSystemService(Context.SENSOR_SERVICE) as SensorManager
    }

    private fun getSensorList() {

        val sensors = sensorManager!!.getSensorList(Sensor.TYPE_ALL)

        for (sensor in sensors) {
            if (!sensorList.contains(sensor))
                sensorList.add(sensor)
        }
        aaSensor!!.notifyDataSetChanged()
    }

}

